### ISSUE TITLE
It is too old. When are you going to update kernel to latest 5.7?

### ISSUE THREAD

**Issue by expnkx** [microsoft/WSL2-Linux-Kernel#95]

**Comment by expnkx** [comment]
I tried to build kernel by myself. The build process is ok, however it is failed to load by WSL2.

**Comment by vanstinator** [comment]
They mentioned in a recent livestream that there were waiting for a few patches to be accepted upstream. Once that happens we'll get 5.7.

**Comment by sileshn** [comment]
https://github.com/nathanchance/WSL2-Linux-Kernel

**Comment by archfrog** [comment]
[Building the WSL 2 kernel with Clang](https://nathanchance.dev/posts/wsl2-kernel-clang/).

In case anybody stumbles across this.

**Comment by expnkx** [comment]
> [Building the WSL 2 kernel with Clang](https://nathanchance.dev/posts/wsl2-kernel-clang/).
> 
> In case anybody stumbles across this.

why use clang for compiling Linux?? Clang is still far away from GCC and the gap is getting larger and larger. clang has the worst C++20 support, for example, even much worse than msvc.

**Comment by archfrog** [comment]
Oh, I guess you're right.  I'm a bit into LLVM so I am a bit biased :-)

The point of the link, IMHO, is that it explains what to do once you've built the kernel.  I just started using WSL seriously today and had no idea how to make WSL2 use my kernel.

**Comment by expnkx** [comment]
> Oh, I guess you're right. I'm a bit into LLVM so I am a bit biased :-)
> 
> The point of the link, IMHO, is that it explains what to do once you've built the kernel. I just started using WSL seriously today and had no idea how to make WSL2 use my kernel.

BTW. GCC can be used on windows directly while clang could not.
Here is my GCC-11 on windows
https://bitbucket.org/ejsvifq_mabmip/mingw-gcc/src/master/

**Comment by expnkx** [comment]
> Also, C++ support shouldn't be an issue for the Linux kernel as far as I know.

Linux has a very long history of using GCC. That is why we called Linux GNU/Linux. All your toolchains are using gcc. glibc, libstdc++

> Chrome is built using Clang, Clang probably has a better and more modern optimizer, and so on.

That is not my experience. For example, clang's std::fstream is horribly slow.

**Comment by archfrog** [comment]
The term "GNU/Linux" was coined by the GNU people, not by the Linux community as such, because the GNU people felt that they were being neglected (which they were).

The Linux kernel doesn't use C++ at all, does it?  I think you're right about Clang's not-yet-quite-mature C++ support, but it is not relevant to the Linux kernel.

But let's not argue needlessly.  I wasn't trying to imply that people should use Clang for building a kernel for WSL2, I was only trying to provide a link that provides some needed background info.

**Comment by expnkx** [comment]
> The Linux kernel doesn't use C++ at all, does it? I think you're right about Clang's not-yet-quite-mature C++ support, but it is not relevant to the Linux kernel.

Linux Kernel is using C89 dialects with many GCC only extensions (like attributes). Normal C++ won't work with it.

**Comment by archfrog** [comment]
Okay, did you try the .wslconfig file trick mentioned in the link I gave?

**Comment by expnkx** [comment]
> Oh, I guess you're right. I'm a bit into LLVM so I am a bit biased :-)
> 
> The point of the link, IMHO, is that it explains what to do once you've built the kernel. I just started using WSL seriously today and had no idea how to make WSL2 use my kernel.

I have put my vmlinux file path into .wslconfig but it does not work. WSL failed to load it. I think this is Microsoft's issue here.

**Comment by expnkx** [comment]
> Okay, did you try the .wslconfig file trick mentioned in the link I gave?
Here was my video about how to build the latest GCC within 20 minutes. You can have a try.
https://www.youtube.com/watch?v=qFToZjo2x24

**Comment by archfrog** [comment]
Thank you!

**Comment by expnkx** [comment]
> https://github.com/nathanchance/WSL2-Linux-Kernel

5.6 is too old. I want 5.7 master

**Comment by nathanchance** [comment]
> why use clang for compiling Linux??

To catch issues that GCC can't:

https://git.kernel.org/linus/79ba4f9310673a8870fda5a062a558923fa65c93
https://git.kernel.org/linus/83c156d3ecc0121d27dc2b7f34e829b265c70c4f
https://git.kernel.org/linus/37209783c73a47692dbf1e6b2dba0d07f6ce24b3
https://git.kernel.org/linus/f7019b7b0ad14bde732b8953161994edfc384953
https://git.kernel.org/linus/e7140639b1de65bba435a6bd772d134901141f86

and in the case of my tree, taking advantage of Google's LTO and CFI implementation.

> Clang is still far away from GCC and the gap is getting larger and larger. clang has the worst C++20 support, for example, even much worse than msvc.

And that's relevant for the kernel how?

> Linux Kernel is using C89 dialects with many GCC only extensions (like many attributes). Normal C/C++ toolchain won't work with it.

What is a "normal C/C++ toolchain"? `clang` supports most of the GCC only extensions and the kernel handles this just fine through the `compiler_attributes.h` header.

> I have put my vmlinux file path into .wslconfig but it does not work. WSL failed to load it. I think this is Microsoft's issue here.

You need to use `bzImage`, rather than `vmlinux`.

> > https://github.com/nathanchance/WSL2-Linux-Kernel
> 
> 5.6 is too old. I want 5.7 master.

Uhhh...

```
$ head -5 Makefile
# SPDX-License-Identifier: GPL-2.0
VERSION = 5
PATCHLEVEL = 7
SUBLEVEL = 0
EXTRAVERSION = -rc5
```

Furthermore, my tree is based on `linux-next`, the integration tree, which has everything slated for 5.8 (or sometimes even 5.9) so it's pretty much as new as it can get.

**Comment by CuriousTommy** [comment]
I hope they also fix https://github.com/microsoft/WSL2-Linux-Kernel/issues/39 when they update the kernel. It would nice for WSL2 users to be able to run Darling (without any workarounds).

> They mentioned in a recent livestream that there were waiting for a few patches to be accepted upstream. Once that happens we'll get 5.7.

@vanstinator Do you have a link to that livestream?

**Comment by tajetaje** [comment]
+1 for fixing #39 as this (to me) seems like a fairly simple implementation tweak that would allow Darling (and some other kernel modules) to function without hacky workarounds.

**Comment by CarterLi** [comment]
> > [Building the WSL 2 kernel with Clang](https://nathanchance.dev/posts/wsl2-kernel-clang/).
> > In case anybody stumbles across this.
> 
> why use clang for compiling Linux?? Clang is still far away from GCC and the gap is getting larger and larger. clang has the worst C++20 support, for example, even much worse than msvc.

No. The major feature that clang lacks is full concept support, others are mostly the same.

And because GCC has much long history than clang, GCC has more historical baggage. For example: https://stackoverflow.com/questions/46901697/is-returning-a-2-tuple-less-efficient-than-stdpair#answer-46901813

If you want to experiment io_uring with C++ coroutines. I'd suggest you use clang because GCC does MUCH worse optimizations for co_await things than clang.

**Comment by expnkx** [comment]
> > > [Building the WSL 2 kernel with Clang](https://nathanchance.dev/posts/wsl2-kernel-clang/).
> > > In case anybody stumbles across this.
> > 
> > 
> > why use clang for compiling Linux?? Clang is still far away from GCC and the gap is getting larger and larger. clang has the worst C++20 support, for example, even much worse than msvc.
> 
> No. The major feature that clang lacks is full concept support, others are mostly the same.
> 
> And because GCC has much long history than clang, GCC has more historical baggage. For example: https://stackoverflow.com/questions/46901697/is-returning-a-2-tuple-less-efficient-than-stdpair#answer-46901813
> 
> If you want to experiment io_uring with C++ coroutines. I'd suggest you use clang because GCC does MUCH worse optimizations for co_await things than clang.

Funny how people clang to be better when it does not even correctly support Concepts. Coroutine is useless compared to Concepts.

**Comment by Biswa96** [comment]
At least there is a 5.4 branch here https://github.com/microsoft/WSL2-Linux-Kernel/tree/linux-msft-5.4.y. Let be happy with it 😁

**Comment by Biswa96** [comment]
There is a 5.4 branch here https://github.com/microsoft/WSL2-Linux-Kernel/tree/linux-msft-5.4.y.

**Comment by levicki** [comment]
Umm... sorry for barging in, but I am curious -- too old for what?

**Comment by martinothamar** [comment]
Personally working a little bit with the io_uring interface, would be convenient if the default kernel was new enough to support that. Currently switching between kernels depending on what I work on. 

I think for most people though having a newer kernel won't make much of a difference so I understand this not being a top priority

**Comment by expnkx** [comment]
> Umm... sorry for barging in, but I am curious -- too old for what?

io_uring too.

WE WANT IO_URING !!!!

**Comment by vicentegarciadiez** [comment]
But, taking into account this branch [https://github.com/microsoft/WSL2-Linux-Kernel/tree/v5.5.6](https://github.com/microsoft/WSL2-Linux-Kernel/tree/v5.5.6)

How can I compile this kernel and start to using it in WSL2?
